primeqa/ir/dense/colbert_top/colbert/search/index_storage.py

```diff
@@ -9,38 +10,72 @@
 from primeqa.ir.dense.colbert_top.colbert.search.candidate_generation import CandidateGeneration
 
 from .index_loader import IndexLoader
-from primeqa.ir.dense.colbert_top.colbert.modeling.colbert import colbert_score, colbert_score_packed
-
-
-class IndexScorer(IndexLoader, CandidateGeneration):
-    def __init__(self, index_path):
-        super().__init__(index_path)
-
-        self.embeddings_strided = ResidualEmbeddingsStrided(self.codec, self.embeddings, self.doclens)
-
-        print_message("#> Building the emb2pid mapping..")
-        all_doclens = load_doclens(index_path, flatten=False)
+from primeqa.ir.dense.colbert_top.colbert.modeling.colbert import colbert_score, colbert_score_packed, colbert_score_reduce
 
-        assert self.num_embeddings == sum(flatten(all_doclens))
+from math import ceil
 
-        all_doclens = flatten(all_doclens)
-        total_num_embeddings = sum(all_doclens)
+import os
+import pathlib
+from torch.utils.cpp_extension import load
+import sys
 
-        self.emb2pid = torch.zeros(total_num_embeddings, dtype=torch.int)
+class IndexScorer(IndexLoader, CandidateGeneration):
+    def __init__(self, index_path, use_gpu):
+        super().__init__(index_path, use_gpu=use_gpu)
 
-        """
-        EVENTUALLY: Build this in advance and load it from disk.
+        IndexScorer.try_load_torch_extensions(use_gpu)
 
-        EVENTUALLY: Use two tensors. emb2pid_offsets will have every 256th element. emb2pid_delta will have the delta
-                    from the corresponding offset, 
-        """
+        self.embeddings_strided = ResidualEmbeddingsStrided(self.codec, self.embeddings, self.doclens)
 
-        offset_doclens = 0
-        for pid, dlength in enumerate(all_doclens):
-            self.emb2pid[offset_doclens: offset_doclens + dlength] = pid
-            offset_doclens += dlength
+    @classmethod
+    def try_load_torch_extensions(cls, use_gpu):
+        if hasattr(cls, "loaded_extensions") or use_gpu:
+            return
+
+        verbose=os.getenv("COLBERT_LOAD_TORCH_EXTENSION_VERBOSE", "False") == "True"
+
+        print_message(f"Loading filter_pids_cpp extension (set COLBERT_LOAD_TORCH_EXTENSION_VERBOSE=True for more info)...")
+        try:
+            filter_pids_cpp = load(
+                name="filter_pids_cpp",
+                sources=[
+                    os.path.join(
+                        pathlib.Path(__file__).parent.resolve(), "filter_pids.cpp"
+                    ),
+                ],
+                extra_cflags=["-O3"],
+                verbose=verbose,
+            )
+        except (RuntimeError, KeyboardInterrupt) as e:
+            if not verbose:
```

Comment:
Printing stack if not verbose?
But, it's the same logic as in 

> modeling/colbert.py:59:            if not verbose:
search/index_storage.py:50:            if not verbose:
search/index_storage.py:70:            if not verbose:
indexing/codecs/residual.py:129:            if not verbose:
indexing/codecs/residual.py:152:            if not verbose: